applications/TrilinosApplication/custom_strategies/convergencecriterias/trilinos_displacement_criteria.h

```diff
@@ -4,332 +4,285 @@
+    void PrintInfo(std::ostream& rOStream) const override
+    {
+        rOStream << Info();
+    }
+
+    /// Print object's data.
+    void PrintData(std::ostream& rOStream) const override
+    {
+        rOStream << Info();
+    }
 
     ///@}
     ///@name Friends
     ///@{
 
-
     ///@}
-
 protected:
     ///@name Protected static Member Variables
     ///@{
 
-
     ///@}
     ///@name Protected member Variables
     ///@{
 
-
     ///@}
     ///@name Protected Operators
     ///@{
 
-
     ///@}
     ///@name Protected Operations
     ///@{
 
+    /**
+     * @brief This method assigns settings to member variables
+     * @param ThisParameters Parameters that are assigned to the member variables
+     */
+    void AssignSettings(const Parameters ThisParameters) override
+    {
+        BaseType::AssignSettings(ThisParameters);
+    }
 
     ///@}
     ///@name Protected  Access
     ///@{
 
-
     ///@}
     ///@name Protected Inquiry
     ///@{
 
-
     ///@}
     ///@name Protected LifeCycle
     ///@{
 
-
-
     ///@}
-
 private:
     ///@name Static Member Variables
     ///@{
 
-
     ///@}
     ///@name Member Variables
     ///@{
-    TDataType mRatioTolerance;
-    TDataType mAlwaysConvergedNorm;
-
-    // Epetra_MpiComm& mrComm;
-
 
-    TDataType mReferenceDispNorm;
     ///@}
     ///@name Private Operators
     ///@{
 
-    TDataType CalculateReferenceNorm(DofsArrayType& rDofSet, ModelPart& rModelPart)
-    {
-        TDataType local_ReferenceDispNorm = TDataType();
-        TDataType value;
-
-        const int rank = rModelPart.GetCommunicator().MyPID();
-
-        for(auto it_dof = rDofSet.begin() ; it_dof != rDofSet.end() ; ++it_dof)
-        {
-            if(it_dof->IsFree() && (it_dof->GetSolutionStepValue(PARTITION_INDEX) == rank))
-            {
-                value = it_dof->GetSolutionStepValue();
-                local_ReferenceDispNorm += value*value;
-            }
-        }
-
-        //perform the sum between all of the nodes
-        TDataType ReferenceDispNorm = local_ReferenceDispNorm;
-        ReferenceDispNorm = rModelPart.GetCommunicator().GetDataCommunicator().SumAll(ReferenceDispNorm);
-
-
-        ReferenceDispNorm = std::sqrt(ReferenceDispNorm);
-        return ReferenceDispNorm;
-    }
-
     ///@}
     ///@name Private Operations
     ///@{
 
+    /**
+     * @brief Check if a Degree of Freedom (Dof) is free.
+     * @details This function checks if a given Degree of Freedom (Dof) is free.
+     * @param rDof The Degree of Freedom to check.
+     * @param Rank The rank of the Dof.
+     * @return True if the Dof is free, false otherwise.
+     * @todo We should doo as in the residual criteria, and consider the active DoFs (not just free), taking into account the MPC in addition to fixed DoFs
+     */
+    bool IsFreeDof(
+        const DofType& rDof,
+        const int Rank
+        ) override
+    {
+        return (rDof.IsFree() && (rDof.GetSolutionStepValue(PARTITION_INDEX) == Rank));
```

Comment:
why do you need to check the PARTITION_INDEX here? the fixity status should be consistent on any Cpu

Reply:
btw, this function is very small, making it virtual makes an important overhead

Reply:
I can avoid virtual with some static polimorfism

Reply:
You must check that it is on the partition, otherwise you are summing several times the same value into the norm

Reply:
See the previous implementation

Reply:
ok

Reply:
just it is a wrong naming for the funciton then. we should have a comment in the code to explain that it is done for that reason

Reply:
Do you mean in the soace function or in the convergence criteria one?